pkg/nimble/netif/nimble_netif.c

```diff
@@ -635,63 +665,164 @@ int nimble_netif_close(int handle)
-                   uint32_t timeout,
-                   const struct ble_gap_adv_params *adv_params)
+#if MYNEWT_VAL_BLE_EXT_ADV
+static int _get_phy_hci(uint8_t mode)
 {
-    assert(adv_params);
+    switch (mode) {
+        case NIMBLE_PHY_1M:
+            return BLE_HCI_LE_PHY_1M;
+#if IS_USED(MODULE_NIMBLE_PHY_2MBIT)
+        case NIMBLE_PHY_2M:
+            return BLE_HCI_LE_PHY_2M;
+#endif
+#if IS_USED(MODULE_NIMBLE_PHY_CODED)
+        case NIMBLE_PHY_CODED:
+            return BLE_HCI_LE_PHY_CODED;
+#endif
+        default:
+            return -1;
+    }
+}
+#endif
 
+static int _accept(const uint8_t *ad, size_t ad_len, const ble_addr_t *addr,
+                   const nimble_netif_accept_cfg_t *params)
+{
     int handle;
     int res;
     (void)res;
 
+    assert(params);
+
     /* allocate a connection context for incoming connections */
     handle = nimble_netif_conn_start_adv();
     if (handle < 0) {
         return handle;
     }
 
-    /* set advertisement data */
-    if (ad != NULL) {
-        res = ble_gap_adv_set_data(ad, (int)ad_len);
-        assert(res == 0);
-    }
     /* remember address if applicable */
     if (addr) {
         nimble_netif_conn_t *conn = nimble_netif_conn_get(handle);
         bluetil_addr_swapped_cp(addr->val, conn->addr);
     }
 
-    /* remember context and start advertising */
-    res = ble_gap_adv_start(nimble_riot_own_addr_type, addr, timeout,
-                            adv_params, _on_gap_slave_evt, (void *)handle);
-    assert(res == 0);
+#if MYNEWT_VAL_BLE_EXT_ADV
+    struct ble_gap_ext_adv_params p;
+    memset(&p, 0, sizeof(p));
 
-    _notify(handle, NIMBLE_NETIF_ACCEPTING, _netif.l2addr);
+    /* figure out PHY modes */
+    int phy_pri = _get_phy_hci(params->primary_phy);
+    int phy_sec = _get_phy_hci(params->secondary_phy);
+    if ((phy_pri < 0) || (phy_sec < 0)) {
+        nimble_netif_conn_free(handle, NULL);
+        return -EINVAL;
+    }
+    /* the 2M PHY is not allowed as primary phy, we need to used the 1M PHY
+     * instead. This is for convenience so uses may define 2M as primary PHY */
+    if (phy_pri == BLE_HCI_LE_PHY_2M) {
+        phy_pri = BLE_HCI_LE_PHY_1M;
+    }
+
+    if (addr != NULL) {
+        p.directed = 1;
+        memcpy(&p.peer, addr, sizeof(p.peer));
+        if (params->flags & NIMBLE_NETIF_FLAG_HD_MODE) {
+            p.high_duty_directed = 1;
+        }
+    }
+    else {
+        p.connectable = 1;
+    }
+    if (params->flags & NIMBLE_NETIF_FLAG_LEGACY) {
+        p.legacy_pdu = 1;
+    }
+    p.itvl_min = BLE_GAP_ADV_ITVL_MS(params->adv_itvl_ms);
+    p.itvl_max = BLE_GAP_ADV_ITVL_MS(params->adv_itvl_ms);
+    p.channel_map = params->channel_map;
+    p.own_addr_type = params->own_addr_type;
+    p.primary_phy = (uint8_t)phy_pri;
+    p.secondary_phy = (uint8_t)phy_sec;
+    p.tx_power = params->tx_power;
+
+    res = ble_gap_ext_adv_configure(EXT_ADV_INST, &p, NULL,
+                                    _on_gap_slave_evt, (void *)handle);
```

Comment:
Doesn't the advertising instance need to be disabled to be able to configure it, and since this will be called multiple times and the instance is never removed then after the first call `configure` will always fail never updating the parameters.

```suggestion
    if (ble_gap_ext_adv_active(EXT_ADV_INST)) {
        res = ble_gap_ext_adv_remove(EXT_ADV_INST);
        assert(rc == 0);
    }
    res = ble_gap_ext_adv_configure(EXT_ADV_INST, &p, NULL,
                                    _on_gap_slave_evt, (void *)handle);
```

Reply:
yes and no. It needs to be disabled to be configured, that is true. But if the code gets to this place it is always disabled, as the accept is either canceled because `nimble_netif_accept_stop()` was called before hand, or because a new connection was established.

If one would call `nimble_accept` or `nimble_accept_direct()` while accepting is in progress, the function would return with an error before event getting to this point here.

Reply:
Ok!